launch_frontend/launch_frontend/parse_substitution.py

```diff
@@ -0,0 +1,108 @@
+# Copyright 2019 Open Source Robotics Foundation, Inc.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+"""Module for parsing substitutions."""
+
+import os
+from typing import Text
+
+from lark import Lark
+from lark import Token
+from lark import Transformer
+
+from launch.substitutions import TextSubstitution
+
+from launch_frontend.expose import substitution_parse_methods
+
+
+def replace_escaped_characters(data: Text, beg: int = 0) -> Text:
+    """Search escaped characters and replace them."""
+    pos = data.find('\\', beg)
+    if pos == -1:
+        return data[beg:]
+    return data[beg:pos] + data[pos+1] + replace_escaped_characters(data, pos+2)
+
+
+class ExtractSubstitution(Transformer):
+    """Extract a substitution."""
+
+    def part(self, content):
+        assert(len(content) == 1)
+        content = content[0]
+        if isinstance(content, Token):
+            assert content.type.endswith('_RSTRING')
+            return TextSubstitution(text=replace_escaped_characters(content.value))
+        return content
+
+    single_quoted_part = part
+    double_quoted_part = part
+
+    def value(self, parts):
+        if len(parts) == 1 and isinstance(parts[0], list):
+            # Deal with single and double quoted templates
+            return parts[0]
+        return parts
+
+    single_quoted_value = value
+    double_quoted_value = value
+
+    def arguments(self, values):
+        if len(values) > 1:
+            # Deal with tail recursive argument parsing
+            return [*values[0], values[1]]
+        return values
+
+    single_quoted_arguments = arguments
+    double_quoted_arguments = arguments
+
+    def substitution(self, args):
+        assert len(args) >= 1
+        name = args[0]
+        assert isinstance(name, Token)
+        assert name.type == 'IDENTIFIER'
+        # TODO(hidmic): Lookup and instantiate Substitution
+        if name.value not in substitution_parse_methods:
```

Comment:
Could you expand your idea? I don't understand the benefits of using an `instantiate_substitution` function (it will finally access the internal data).

Reply:
It will access data structures eventually, yes, but you stabilize an API in case anything changes in the future (e.g. how we instantiate, how lookup, what the data structure is, etc.).

Reply:
Still applies.

Reply:
I'm fine with bringing another reviewer but I still think this applies.